File: HtOMSWidget/SalesBoardWidget.swift
import SwiftUI
import WidgetKit

struct SalesBoardEntry: TimelineEntry {
    let date: Date
    let snapshot: DashboardWidgetSnapshot
}

struct SalesBoardProvider: TimelineProvider {
    func placeholder(in context: Context) -> SalesBoardEntry {
        SalesBoardEntry(date: .now, snapshot: .sample)
    }

    func getSnapshot(in context: Context, completion: @escaping (SalesBoardEntry) -> Void) {
        completion(SalesBoardEntry(date: .now, snapshot: storedSnapshot))
    }

    func getTimeline(in context: Context, completion: @escaping (Timeline<SalesBoardEntry>) -> Void) {
        let entry = SalesBoardEntry(date: .now, snapshot: storedSnapshot)
        let nextRefresh = Calendar.current.date(byAdding: .minute, value: 15, to: .now) ?? .now
        completion(Timeline(entries: [entry], policy: .after(nextRefresh)))
    }

    private var storedSnapshot: DashboardWidgetSnapshot {
        WidgetSnapshotStore().load() ?? .sample
    }
}

struct SalesBoardWidget: Widget {
    private let kind = "SalesBoardWidget"

    var body: some WidgetConfiguration {
        StaticConfiguration(kind: kind, provider: SalesBoardProvider()) { entry in
            SalesBoardWidgetView(entry: entry)
        }
        .configurationDisplayName("한통 OMS 현황판")
        .description("오늘과 월간 매출 현황을 공항 안내판처럼 표시합니다.")
        .supportedFamilies([.systemSmall, .systemMedium, .systemLarge])
        .contentMarginsDisabled()
    }
}

private struct SalesBoardWidgetView: View {
    @Environment(\.widgetFamily) private var family
    let entry: SalesBoardEntry

    var body: some View {
        Group {
            switch family {
            case .systemLarge:
                largeBoard
            case .systemSmall:
                smallBoard
            default:
                mediumBoard
            }
        }
        .containerBackground(for: .widget) { BrandPalette.background }
    }

    private var smallBoard: some View {
        VStack(alignment: .leading, spacing: 0) {
            compactValueRow("금일(\(currentDay))", value: entry.snapshot.todayAmount)
            compactValueRow("전일", value: entry.snapshot.yesterdayAmount)
            compactValueRow("당월(\(currentMonth))", value: entry.snapshot.monthTotal)
            compactValueRow("전월", value: entry.snapshot.previousMonthTotal)
            compactValueRow("평균", value: entry.snapshot.dailyAverage)
            compactValueRow("목표(\(achievementRate))", value: entry.snapshot.targetAmount)
            compactTimestampRow("시간", date: entry.snapshot.refreshedAt)
            compactTimestampRow("서버", date: entry.snapshot.serverTime)
        }
        .padding(10)
    }

    private var mediumBoard: some View {
        VStack(alignment: .leading, spacing: 10) {
            boardHeader
            HStack(alignment: .top, spacing: 12) {
                VStack(spacing: 7) {
                    valueRow("금일", value: entry.snapshot.todayAmount)
                    valueRow("전일", value: entry.snapshot.yesterdayAmount)
                }
                VStack(spacing: 7) {
                    valueRow("당월", value: entry.snapshot.monthTotal)
                    valueRow("목표", value: entry.snapshot.targetAmount)
                }
            }
            refreshFooter
        }
        .padding(16)
    }

    private var largeBoard: some View {
        VStack(alignment: .leading, spacing: 9) {
            boardHeader
            valueRow("금일", value: entry.snapshot.todayAmount)
            valueRow("전일", value: entry.snapshot.yesterdayAmount)
            valueRow("당월", value: entry.snapshot.monthTotal)
            valueRow("전월", value: entry.snapshot.previousMonthTotal)
            valueRow("평균", value: entry.snapshot.dailyAverage)
            valueRow("목표", value: entry.snapshot.targetAmount)
            Spacer(minLength: 0)
            timestampRow("갱신", date: entry.snapshot.refreshedAt)
            timestampRow("서버", date: entry.snapshot.serverTime)
        }
        .padding(18)
    }

    private var boardHeader: some View {
        HStack(spacing: 8) {
            Text("현황판")
                .font(.system(.headline, design: .monospaced).weight(.bold))
                .foregroundStyle(.white)
            Text("매출 · 만원")
                .font(.system(.caption, design: .monospaced))
                .foregroundStyle(BrandPalette.mutedText)
            Spacer()
            if entry.snapshot.isSample {
                Text("SAMPLE")
                    .font(.system(.caption2, design: .monospaced).weight(.bold))
                    .foregroundStyle(BrandPalette.orange)
            }
        }
    }

    private func valueRow(_ label: String, value: Int) -> some View {
        HStack(spacing: 8) {
            Text(label)
                .foregroundStyle(BrandPalette.mutedText)
            Spacer(minLength: 6)
            Text(compactWon(value))
                .foregroundStyle(.white)
                .monospacedDigit()
                .privacySensitive()
        }
        .font(.system(.subheadline, design: .monospaced).weight(.semibold))
        .padding(.horizontal, 9)
        .padding(.vertical, 6)
        .background(BrandPalette.boardCell, in: RoundedRectangle(cornerRadius: 6))
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(label) 매출 \(value.formatted())원")
    }

    private func compactValueRow(_ label: String, value: Int) -> some View {
        HStack(spacing: 2) {
            Text(label)
                .foregroundStyle(BrandPalette.mutedText)
                .font(.system(size: 15, weight: .bold, design: .monospaced))
                .lineLimit(1)
                .fixedSize(horizontal: true, vertical: false)
                .layoutPriority(1)
            Spacer(minLength: 1)
            Text(compactWon(value))
                .foregroundStyle(.white)
                .font(.system(size: 17, weight: .bold, design: .monospaced))
                .monospacedDigit()
                .privacySensitive()
        }
        .lineLimit(1)
        .minimumScaleFactor(0.9)
    }

    private func compactTimestampRow(_ label: String, date: Date) -> some View {
        HStack(spacing: 4) {
            Text(label)
                .foregroundStyle(BrandPalette.mutedText)
            Spacer(minLength: 2)
            Text(compactTimestamp(date))
                .foregroundStyle(.white)
                .monospacedDigit()
        }
        .font(.system(size: 12, weight: .semibold, design: .monospaced))
        .lineLimit(1)
        .minimumScaleFactor(0.9)
    }

    private var refreshFooter: some View {
        HStack {
            Text("갱신")
            Spacer()
            Text(entry.snapshot.refreshedAt, format: .dateTime.month().day().hour().minute())
        }
        .font(.system(.caption2, design: .monospaced))
        .foregroundStyle(BrandPalette.mutedText)
    }

    private func timestampRow(_ label: String, date: Date) -> some View {
        HStack {
            Text(label)
            Spacer()
            Text(date, format: .dateTime.month().day().hour().minute())
        }
        .font(.system(.caption, design: .monospaced))
        .foregroundStyle(BrandPalette.mutedText)
    }

    private func compactWon(_ value: Int) -> String {
        Int((Double(value) / 10_000).rounded()).formatted(.number.grouping(.automatic))
    }

    private var currentDay: Int {
        Calendar.current.component(.day, from: entry.snapshot.refreshedAt)
    }

    private var currentMonth: String {
        String(format: "%02d", Calendar.current.component(.month, from: entry.snapshot.refreshedAt))
    }

    private var achievementRate: Int {
        guard entry.snapshot.targetAmount > 0 else { return 0 }
        return Int((Double(entry.snapshot.monthTotal) / Double(entry.snapshot.targetAmount) * 100).rounded())
    }

    private func compactTimestamp(_ date: Date) -> String {
        let parts = Calendar.current.dateComponents([.month, .day, .hour, .minute], from: date)
        return String(
            format: "%02d-%02d %02d:%02d",
            parts.month ?? 0,
            parts.day ?? 0,
            parts.hour ?? 0,
            parts.minute ?? 0
        )
    }
}

#Preview(as: .systemMedium) {
    SalesBoardWidget()
} timeline: {
    SalesBoardEntry(date: .now, snapshot: .sample)
}

#Preview(as: .systemSmall) {
    SalesBoardWidget()
} timeline: {
    SalesBoardEntry(date: .now, snapshot: .sample)
}

#Preview(as: .systemLarge) {
    SalesBoardWidget()
} timeline: {
    SalesBoardEntry(date: .now, snapshot: .sample)
}

File: app/src/main/java/com/htoms/brief/widget/SalesBoardWidget.kt
package com.htoms.brief.widget

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.DpSize
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.glance.GlanceId
import androidx.glance.GlanceModifier
import androidx.glance.LocalSize
import androidx.glance.appwidget.GlanceAppWidget
import androidx.glance.appwidget.GlanceAppWidgetReceiver
import androidx.glance.appwidget.SizeMode
import androidx.glance.appwidget.cornerRadius
import androidx.glance.appwidget.provideContent
import androidx.glance.background
import androidx.glance.layout.Alignment
import androidx.glance.layout.Column
import androidx.glance.layout.Row
import androidx.glance.layout.Spacer
import androidx.glance.layout.fillMaxSize
import androidx.glance.layout.fillMaxWidth
import androidx.glance.layout.height
import androidx.glance.layout.padding
import androidx.glance.layout.width
import androidx.glance.text.FontFamily
import androidx.glance.text.FontWeight
import androidx.glance.text.Text
import androidx.glance.text.TextStyle
import androidx.glance.unit.ColorProvider
import com.htoms.brief.model.DashboardWidgetSnapshot

/** 홈 화면 위젯 리시버. 시스템 APPWIDGET_UPDATE를 Glance 위젯으로 전달한다. */
class SalesBoardGlanceWidgetReceiver : GlanceAppWidgetReceiver() {
    override val glanceAppWidget: GlanceAppWidget = SalesBoardWidget()
}

/**
 * 8행 공항 안내판 스타일 매출 현황 위젯.
 * 앱이 저장한 표시 전용 스냅샷만 읽으며, 없으면 SAMPLE 데이터를 보여 준다.
 */
class SalesBoardWidget : GlanceAppWidget() {

    override val sizeMode: SizeMode = SizeMode.Responsive(
        setOf(SMALL, MEDIUM, LARGE)
    )

    override suspend fun provideGlance(context: Context, id: GlanceId) {
        val snapshot = WidgetSnapshotStore(context).load() ?: DashboardWidgetSnapshot.sample
        provideContent {
            BoardContent(snapshot)
        }
    }

    @Composable
    private fun BoardContent(snapshot: DashboardWidgetSnapshot) {
        val size = LocalSize.current
        Column(
            modifier = GlanceModifier
                .fillMaxSize()
                .background(BACKGROUND)
        ) {
            when {
                size.height >= LARGE.height && size.width >= LARGE.width -> LargeBoard(snapshot)
                size.width >= MEDIUM.width -> MediumBoard(snapshot)
                else -> SmallBoard(snapshot)
            }
        }
    }

    @Composable
    private fun SmallBoard(snapshot: DashboardWidgetSnapshot) {
        Column(modifier = GlanceModifier.fillMaxSize().padding(10.dp)) {
            CompactValueRow("금일(${WidgetFormat.currentDay(snapshot)})", snapshot.todayAmount)
            CompactValueRow("전일", snapshot.yesterdayAmount)
            CompactValueRow("당월(${WidgetFormat.currentMonth(snapshot)})", snapshot.monthTotal)
            CompactValueRow("전월", snapshot.previousMonthTotal)
            CompactValueRow("평균", snapshot.dailyAverage)
            CompactValueRow("목표(${WidgetFormat.achievementRate(snapshot)}%)", snapshot.targetAmount)
            CompactTimestampRow("시간", snapshot.refreshedAt)
            CompactTimestampRow("서버", snapshot.serverTime)
        }
    }

    @Composable
    private fun MediumBoard(snapshot: DashboardWidgetSnapshot) {
        Column(modifier = GlanceModifier.fillMaxSize().padding(16.dp)) {
            BoardHeader(snapshot)
            Spacer(GlanceModifier.height(10.dp))
            Row(modifier = GlanceModifier.fillMaxWidth()) {
                Column(modifier = GlanceModifier.width(90.dp)) {
                    ValueRow("금일", snapshot.todayAmount)
                    Spacer(GlanceModifier.height(7.dp))
                    ValueRow("전일", snapshot.yesterdayAmount)
                }
                Spacer(GlanceModifier.width(8.dp))
                Column(modifier = GlanceModifier.width(90.dp)) {
                    ValueRow("당월", snapshot.monthTotal)
                    Spacer(GlanceModifier.height(7.dp))
                    ValueRow("목표", snapshot.targetAmount)
                }
            }
            Spacer(GlanceModifier.height(4.dp))
            TimestampRow("갱신", snapshot.refreshedAt)
        }
    }

    @Composable
    private fun LargeBoard(snapshot: DashboardWidgetSnapshot) {
        Column(modifier = GlanceModifier.fillMaxSize().padding(18.dp)) {
            BoardHeader(snapshot)
            Spacer(GlanceModifier.height(9.dp))
            listOf(
                "금일" to snapshot.todayAmount,
                "전일" to snapshot.yesterdayAmount,
                "당월" to snapshot.monthTotal,
                "전월" to snapshot.previousMonthTotal,
                "평균" to snapshot.dailyAverage,
                "목표" to snapshot.targetAmount
            ).forEach { (label, value) ->
                ValueRow(label, value)
                Spacer(GlanceModifier.height(9.dp))
            }
            Spacer(GlanceModifier.height(4.dp))
            TimestampRow("갱신", snapshot.refreshedAt)
            TimestampRow("서버", snapshot.serverTime)
        }
    }

    @Composable
    private fun BoardHeader(snapshot: DashboardWidgetSnapshot) {
        Row(
            modifier = GlanceModifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = "현황판",
                style = TextStyle(
                    color = WHITE,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = FontFamily.Monospace
                )
            )
            Text(
                text = " 매출 · 만원",
                style = TextStyle(
                    color = MUTED,
                    fontSize = 11.sp,
                    fontFamily = FontFamily.Monospace
                )
            )
            Spacer(GlanceModifier.width(8.dp))
            if (snapshot.isSample) {
                Text(
                    text = "SAMPLE",
                    style = TextStyle(
                        color = ORANGE,
                        fontSize = 10.sp,
                        fontWeight = FontWeight.Bold,
                        fontFamily = FontFamily.Monospace
                    )
                )
            }
        }
    }

    @Composable
    private fun ValueRow(label: String, value: Long) {
        Row(
            modifier = GlanceModifier
                .fillMaxWidth()
                .background(BOARD_CELL)
                .cornerRadius(6.dp)
                .padding(horizontal = 9.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = label,
                style = TextStyle(
                    color = MUTED,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Medium,
                    fontFamily = FontFamily.Monospace
                )
            )
            Spacer(GlanceModifier.width(12.dp))
            Text(
                text = WidgetFormat.compactWon(value),
                style = TextStyle(
                    color = WHITE,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = FontFamily.Monospace
                )
            )
        }
    }

    @Composable
    private fun CompactValueRow(label: String, value: Long) {
        Row(
            modifier = GlanceModifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = label,
                style = TextStyle(
                    color = MUTED,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = FontFamily.Monospace
                ),
                maxLines = 1
            )
            Spacer(GlanceModifier.width(4.dp))
            Text(
                text = WidgetFormat.compactWon(value),
                style = TextStyle(
                    color = WHITE,
                    fontSize = 15.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = FontFamily.Monospace
                ),
                maxLines = 1
            )
        }
    }

    @Composable
    private fun CompactTimestampRow(label: String, epochMillis: Long) {
        Row(
            modifier = GlanceModifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = label,
                style = TextStyle(
                    color = MUTED,
                    fontSize = 11.sp,
                    fontWeight = FontWeight.Medium,
                    fontFamily = FontFamily.Monospace
                ),
                maxLines = 1
            )
            Spacer(GlanceModifier.width(4.dp))
            Text(
                text = WidgetFormat.compactTimestamp(epochMillis),
                style = TextStyle(
                    color = WHITE,
                    fontSize = 11.sp,
                    fontWeight = FontWeight.Medium,
                    fontFamily = FontFamily.Monospace
                ),
                maxLines = 1
            )
        }
    }

    @Composable
    private fun TimestampRow(label: String, epochMillis: Long) {
        Row(
            modifier = GlanceModifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = label,
                style = TextStyle(color = MUTED, fontSize = 11.sp, fontFamily = FontFamily.Monospace)
            )
            Spacer(GlanceModifier.width(8.dp))
            Text(
                text = WidgetFormat.compactTimestamp(epochMillis),
                style = TextStyle(color = MUTED, fontSize = 11.sp, fontFamily = FontFamily.Monospace)
            )
        }
    }

    companion object {
        private val SMALL = DpSize(110.dp, 110.dp)
        private val MEDIUM = DpSize(220.dp, 110.dp)
        private val LARGE = DpSize(220.dp, 240.dp)

        private val BACKGROUND = ColorProvider(Color(0xFF0E0E0F))
        private val BOARD_CELL = ColorProvider(Color(0xFF18181B))
        private val ORANGE = ColorProvider(Color(0xFFF26436))
        private val MUTED = ColorProvider(Color(0xFFADADB3))
        private val WHITE = ColorProvider(Color.White)
    }
}
